### DISCUSSION TITLE
V4 DISPLAY

### DISCUSSION THREAD

**Discussion by OrtnerMichael** [magpylib/magpylib#394] [General]
# Display

All `display()` features/arguments should be discussed here to achieve maximal user friendliness and feature parity between Matplotlib and Plotly backends.

# API

## Hierarchy of arguments when plotting

**3 - Config (LOWEST)**
`Config.XXXXX`  (LOWEST)

**2 - Individual styles**
style- dict supports underscore_magic
`object.style= style-dict` 
`object(... style=style-dict)`
`object.style.color='red'`

**2 - Collection**
`Collection.set_styles(style-dict and/or style-underscore_magic)`

**0 - explicit (HIGHEST)**
`display(arguments)`
`display(style=style-dict/underscore_magic)`
`display(**style-dict/underscore_magic)`

## Display styling defaults (currently under `magpylib.defaults.display.styles)

Default settings for all plotting values: 
- [x] finalize api

## Display
`display()` arguments globally overwrite default values.

```
magpylib.display(
   *objects, 
   path=True,      #bool, int, index list, 'animate'
   zoom=1,
   animate_time=5,
   markers=None, 
   backend=None,
   canvas=None
   **kwargs)
```

**Comment by OrtnerMichael** [comment]
# Markers
Allow user to display markers in the plot at designated positions. Its a practical feature to mark certain fixed positions in the 3D plot. Specifically practical for debugging.

**Arguments and behavior**
`markers=None`
- `=None` display no markers.
- `=array_like, shape (n,3)` displays markers at all n positions.

`size_markers=1`
- `=float f` adjust marker size with multiplicative argument `f `.

**Matplotlib**
Uses `plt.plot(ls='', marker='o')`

**Plotly**
tbd

**Discussion / 2do**
- [x] Markers is `None` by default
 - [x] Implement `markers` styling in **matplotlib**
 - [x] Implement `markers` styling in **plotly**

**Comment by OrtnerMichael** [comment]
# Plotting Canvas
Choose a canvas (`mpl...Axes`, `plotly...Figure`) where the plot is displayed.

**Arguments and behavior**
`canvas=None`
- `=None` create stand alone canvas and displays immediately
- `=mpl...Axes` or `plotly...Figure` object, creates plot in given object

**Discussion / 2do**
- [x] Implement with **matplotlib**
- [x] Implement with **plotly**

**Reply by Alexboiboi** [reply]
has been adressed and implemented here f8635e850cfdb50d0085a630a3a0967fd40eed3b

**Comment by OrtnerMichael** [comment]
# Direction with arrows/object coloring

Object coloring is automatic. Paths and magnetization vectors should be colored in the respective object color. 
Provide an option to display the magnetization vector and current direction.

**Arguments and behavior**
`show_direction=True`
- `=True`: display direction indication (vectors or coloring).
- `=False`: do not display direction indication. Coloring should be random and homogeneous.

`size_direction=1`
- `=float f` Adjust size of direction vectors with a multiplicative factor `f`.

**Matplotlib**
Uses `plt.quiver()` to display arrows in magnetization direction for magnets.

**Plotly**
Uses magnet coloring when `show_direction=True` with DIN SPEC 14611 coloring by default.
Use Config for 3-color and 2-color options.

**Discussion / 2do**
- [x] Plotly `show_direction=False` should result in differently colored objects like in Matplotlib.
- [x] Make sure that standard physics North-South poles correspond to correct Red-Green coloring.
- [x] set default `show_direction=True`
- [x] missing current direction for Matplotlib

**Reply by Alexboiboi** [reply]
I don't a reason to remove the color arguments in the display function. They don't need to be explicitly shown as named arguments and could simply be set through kwargs

**Reply by Alexboiboi** [reply]
we could have a property argument as a dictionary to access individual object properties. Something like

```python
import magpylib as magpy

mag1 = magpy.magnet.Cuboid((0,0,1), (1,1,1), (0, 0,1))
mag2 = magpy.magnet.Cuboid((1,0,0), (1,1,1), (0,0,-1))

magpy.display(mag1, mag2, properties={mag1: {'color':'blue'}, mag2: {'color':'red'}})
```

**Comment by OrtnerMichael** [comment]
# Show paths
Allow users to see object path positions and orientations.

**Arguments and behavior**
`show_path=True`
 - `=False`: Displays only last path position
 - `=True`: Show the object paths as lines. Display objects at end of path.
 - `=int i`: Show paths as lines. Display objects at every i'th path position
 - `=array_like, shape (n,)`: Show path as lines. Display objects on path positions with given input indices.
 - `='animate'`: Create an animation of the objects moving along the paths.

`animate_time=3`
- `=float t` animation time in seconds

`animate_fps=30`
- `=int n` maximal number of frames per second. displays path positions only.


**Discussion / 2do**
- [x] Matplotlib `show_path=[i1,i2,i3,...]` implementation
- [x] Matplotlib: if `show_path='animate'` default to `True`. Raise warning.
- [x] Config should have a `Config.ANIMATE_MAX_FRAMES=x`
- [x] Plotly adjust argument names
- [x] Show path endpoints when indices beyond path length

**Comment by OrtnerMichael** [comment]
# Sensors
Sensors are displayed with tri-color coordinate crosses at their position, with markers at their pixels and with a sensor-box surrounding the pixel.

**Arguments and behavior**
`size_sensors=1`
- `=float f` Adjust sensor coordinate cross with multiplicative factor `f`
- if `=0` do not display coordinate cross.

`size_pixels=1`
- `=float f` Adjust sensor pixel-marker with multiplicative factor `f`
- if `=0` do not display pixels

**Matplotlib**
Coordinate crosses are displayed with `plt.quiver()`. Pixels with `plt.plot(ls='', marker='o')`.

**Plotly**
Always displays coordinate cross,
Always displays sensor box,
Displays pixels

**Discussion / 2do**
- [x] add `size_pixels` argument
- [x] Matplotlib implement `size_pixels` option
- [x] Introduce size=0 options in Matplotlib
- [x] Plotly reduce auto-pixel size by factor of 2

**Comment by OrtnerMichael** [comment]
# Display window and zoom
Automatic display window size and convenient zoom-option

**Arguments and behavior**
`zoom=1`
- `=float f` Adjust automatic zoom level with multiplicative factor `f`.

**Discussion / 2do**
- [x] Plotly should we always have a Cube-display window ?
- [x] Matplotlib `zoom` default should be 1
- [x] Do we zoom in the right direction ? `zoom=2` means double or half the window size ?

**Comment by Alexboiboi** [comment]
# Magic underscore notation for display function

The amount of arguments in the display function is getting significantly long and since arguments can be grouped, it may be of interest to implement some _magic underscore notation_ to make it cleaner to read and write for the user. Also `poltly` is already using this feature throughout the entire library and the notation system could be used in conjunction with the display function arguments, at least for the plotly backend.

See here how it functions in plotly 
https://plotly.com/python/creating-and-updating-figures/#magic-underscore-notation

The current argument for the display function are:
```python
markers, canvas, show_direction, show_path, size_sensors, size_direction, size_dipoles, 
size_pixels, zoom, plotting_backend, renderer, animate_time, animate_fps, backtofirst,
color_discrete_sequence, show_path_numbering, color_transition, north_color, 
middle_color, south_color, pixel_color, opacity, color
```

and this could be a possible grouping to use the magic undescore with:

```python
- canvas
    - source #(fig or axis)
    - zoom # float (relative), 0 = tight_fit
    - renderer # only for plotly https://plotly.com/python/renderers/
- markers
    - positions # (N,3) array 
    - size # float (absolute, not affected by interactive zoom)
    - color # css color
    - symbol # 'x', 'o' etc
- sensors
    - size # float (relative to canvas)
    - pixel
        - size # float (relative to )
        - color # css color
- dipoles
    - size # relative to canvas
    - direction # show as color or uniform?
- path
    - show # 'animate', int, array
    - time # float, animation time
    - fps # int, animation max frame rate
    - lastframe # int, which frame to stop on after animation finished
    - line # this already a plotly arg
        -> line properties
    - marker # this already a plotly arg
        -> markers properties
- magnetization
    - direction # bool or float relative arrow size
    - color
        - transition # float, between O and 1 for smoothness
        - north # css color
        - middle # css color or False to hide middle color or 'auto' to cycle through
        - south # css color
        - pixel # css color
```

possible use would be something like:

```python
display(
    collection,
    markers=[(0,1,2)], 
    makers_size=2,  # magic underscore
    sensors_size=2, # magic underscore
    sensors_pixel_size=1, # magic underscore
    magnetization=dict(direction=True, middle_color=False) # or simple dict
    # and so on
)
```

**Reply by OrtnerMichael** [reply]
the magic underscore notation is nice but is missing the tool-tip option - and we do not have so many parameters that it is necessary to make use of this. Also there is the danger that we use underscore notation elsewhere for args - this might lead to confusion.

**Comment by Alexboiboi** [comment]
> All `display()` features/arguments should be discussed here to achieve maximal user friendliness and feature parity between Matplotlib and Plotly backends.
> 
> **To discuss**
> 
> * [x]  `size` options as `auto` and absolute size values OR relative with `auto=1`

I think it is ok to leave the default value to 1 as a relative size to the canvas.

I also implemented a new `Config.AUTOSIZE_FACTOR`  which tells the default multiplication factor between  the canvas range and the autosized object. For now it defaults to 10.

example:

`Config.AUTOSIZE_FACTOR` = 10
`size_sensors=2`

`effective sensor size` =  `canvas_size / Config.AUTOSIZE_FACTOR * size_sensor` = `canvas_size/5` 

- Case 1: `Sensor` without pixel, or `Dipole` : 
size is based on the relative `size_sensors`, Config.AUTOSIZE_FACTOR and `canvas_size`

- Case 2: `Sensor` with pixel:
  - size is base on max(pixel envelope size, autosize) to avoid the arrows to disappear into the pixel envelope 

Notes: 
- functions the same in matplotlib except for Case 2. But in this case the sensor CS is still visible anyway. So no changes really needed there. (42f65cc2fef45361d8651162e7e26af55833c910)
- I also fixed the matplotlib representation when only a single dipole is shown. In this case the ranges calculation were raising an error. (feae8c0d244ba3644bc74632c0b2086cf53bfdd4)

**Comment by OrtnerMichael** [comment]
## The object style argument

All magpylib objects should have a new `style=None` **(better name ?)** argument that can be used to set individual object styles. The code should work in several ways:
```
mag = magpy.magnet.Cuboid(...., style={'color':(r,g,b)})
mag.style.color=(r,g,b)
```

Collection styles overwrite individual styles.

`None` defaults to the respective default styles.


- [x] Implement style property for all sources
- [x] object styles apply in plotly
- [x] object styles apply in matplotlib
- [x] color currents in matplotlib

**Reply by Alexboiboi** [reply]
I already made draft which works with color here 20edb7e7d07e74cb7614e0a203043f10f5ae63b9

**Comment by OrtnerMichael** [comment]
# BUG 001

View-window in Matplotlib not properly centered
```
import magpylib as magpy
col = magpy.Collection([magpy.magnet.Cuboid((0,0,1), (1,1,1), (i,0,0)) for i in range(10)])
col.display()
```

<img src="https://user-images.githubusercontent.com/47384012/139557520-a6a44bf5-8485-444e-859b-0a7835f6625f.png" width=20% height=20%>

**Reply by Alexboiboi** [reply]
<img src="https://user-images.githubusercontent.com/29252289/139558522-d59981c4-efc9-4794-a765-7126933946ce.png" width=20% height=20%>
fixed in d1ed7b3b8a547ce2035a5c6cba877eeb51986586

**Reply by OrtnerMichael** [reply]
this also fixed the viewwindow-not-cubical bug :)

**Comment by OrtnerMichael** [comment]
# BUG 002

Some individual styles are not recognized (e.g. "x has no property 'size'") 
```
current.style.size=10
```

Many individual styles do not show in tool tips
```
current.style.current
sphere.style.magnetization
dipole.style.magnetization
```

i have to make a systematic study to test them all - then I'll update

**Reply by Alexboiboi** [reply]
> current.style.size=10

If `current` is an instance of one of `Circular` or `Line`, the correct syntax is:

```python
current.style.current.size=10
```

the reason is to avoid setting size of both dipoles and current when we do this:

```python
display(dipole, line, style_size=4)
```

the syntax in now:

```python
display(dipole, line, style_current_size=4)
```
Otherwise I would need to introduce special case in the syntax parsing the kwargs...

**Reply by OrtnerMichael** [reply]
richtig macht  auch Sinn so

**Comment by Alexboiboi** [comment]
Hi @OrtnerMichael 

I forked the _plotly_dev_ branch again to try out a new way of implementing the default style values.  

you can find it here: https://github.com/magpylib/magpylib/tree/style_parampyviz

I basically removed all the config parameter related to specific object styling and implemented a new base styling class `MagpylibStyle` containing the default styling options by family:
- Base (common for all families)
- Magnets
- Currents
- Dipoles
- Sensors

the defaults can be set with the _dot_ notation, or a dictionary (including magic underscore) and validators checks user inputs at every level. The `style` attributes points at an instance of the `MagpylibStyle` default class and sits at the library top level. Also all default values can be reset.

- [ ]  implement reset method for every subclass

Here a short example of how it could be used

```python
import magpylib as magpy
magpy.style.base.path.marker.size = 10
#magpy.style.reset()

mag1 = (
    magpy.magnet.Cuboid(magnetization=(1,0,0), dimension=(1, 1, 1), position=(3, 0, 3))
    .rotate_from_angax([1] * 100, "x", anchor=(0, 0, 6), increment=True)
).display()
```

**Reply by OrtnerMichael** [reply]
looks cool - I made it a discussion point below

**Comment by OrtnerMichael** [comment]
# BUG 003

`obj.style.opacity=x` is not applied
also not at top-level (in display())

**Reply by Alexboiboi** [reply]
fixed in 39ac15907b03338f4b409cae5763ea578c340178

**Comment by OrtnerMichael** [comment]
# BUG 004

Matplotlib Dipole is an arrow that starts at dipole.position - however the center of the arrow should be there (like in Plotly)

see also image in BUG 006 below

*Jammern auf hohem niveau*

**Reply by Alexboiboi** [reply]
Implement a pivot styling argument:

```python
class Dipoles(BaseProperties):
    """
    Defines the specific styling properties of the objects of the `dipoles` family

    Properties
    ----------
    size: float, default=None
        positive float for relative dipole to size to canvas size

    pivot: str, default=None
        the part of the arrow that is anchored to the X, Y grid.
        The arrow rotates about this point. Can be one of `['tail', 'middle', 'tip']`
 """
```

**Reply by OrtnerMichael** [reply]
now its only in the middle

`style_dipole_pivot='tip' ` doesn't do anything

**Reply by Alexboiboi** [reply]
> `style_dipole_pivot='tip' ` doesn't do anything

the correct syntax is `style_pivot='tip'`

**Comment by OrtnerMichael** [comment]
# BUG 005

Dipole now has `style.magnetization` attribute and shows a "magnetization".
This should not be the case for 2 reasons:
1. The arrow alone is enough to indicate the moment direction
2. A dipole does not physically have a magnetization

`Dipole_object.style` should not have the `.magnetization` method and should also not display the magnetization color-pattern.

**Reply by Alexboiboi** [reply]
fixed in 9c4b45821b94a2eddf8a1ec5f5a275900c0526d1

**Comment by OrtnerMichael** [comment]
# Base-Style suggestions

- [x] Current base-linewidth appears much fatter in Plotly than in Matplotlib (looks better in Plotly)
- [x] Path-points appear much bigger in Matplotlib than in Plotly. Plotly path markers are not visible.
- [x] Dipoles appear much bigger in Plotly than in Matplotlib
- [x] Default zoom level seems a bit far out - maybe we should use `zoom=0` by default ?

**Reply by Alexboiboi** [reply]
for the zoom level , i am i favor of having `zoom=0`as default

for the rest i think we need a live discussion

**Comment by OrtnerMichael** [comment]
# For Discussion
- [x] Is there a possibility to allow the much-used Matplotlib standard colors allow matplotlib colors 'r' 'g' 'b' 'y' 'm' 'c' 'k' ?
- [x] Cannot turn on/off individual paths with styling
- [x] Marker style not in config

**Reply by Alexboiboi** [reply]
> *  Cannot turn on/off individual paths with styling

It is no problem to implement but I am not sure why you would want this on an individual level. It could get quite confusing if you start show only some paths. How do you differentiate a path that is hidden, with a object which is static then? 
Also it is not clear how to handle this feature since the `path` argument is not part of the styling ecosystem, so we would need to define precedence again if `path=True` conficts with `style_path_show=True`

**Reply by Alexboiboi** [reply]
> 'r' 'g' 'b' 'y' 'm' 'c' 'k' 

- [x] Allow short color names for styling

done in 0549e5b615caaba618428b158fa8c09f7407ef2c

**Reply by Alexboiboi** [reply]
>  Cannot turn on/off individual paths with styling

it is actually possible indirectly with:
‘’’python
path.marker.size = 0
path.line.width = 0
‘’’

**Comment by OrtnerMichael** [comment]
# BUG 002 part2

Individual styling methods `.style.current`, `.style.magnetization` and `.style.pixel` and `.style.size` are not seen in tool tips

**Reply by Alexboiboi** [reply]
I don't find any explanation on why this is so. Therefore it is not possible for me to fix this for now. 
These attributes are accessible though..

**Reply by OrtnerMichael** [reply]
maybe @lucasgcb has an idea how to solve this problem ?

How to inform a code inspector that a class has certain methods/members that the inspector doesnt find by himself ?

**Comment by OrtnerMichael** [comment]
# Docstrings

`.update_styles()` and `display()` docstring
- [ ] must explain style input format
- [ ] must have examples

maybe this comes last when everything is working - but difficult for me to know the input options :)

**Comment by OrtnerMichael** [comment]
# BUG 006

path-marker in Plotly are always black - in Matplotlib thy have the path color

![image](https://user-images.githubusercontent.com/47384012/139581745-b680ed34-2a1f-4f9a-864f-ef1c893f9dc0.png)

**Reply by Alexboiboi** [reply]
The  path markers and the path lines now take the object color unless a color is explicitly set for them for example lie :
`style.path.marker.color =  'some_color'`
- [x] fix markers behavior

**Comment by OrtnerMichael** [comment]
# Config Discussion

- [x] config structure
- [x] config tool tips
- [x] moving styles from `magpy.Config` to `magpylib.defauls.display.style`
- [x] Hidden Config ?

**Reply by Alexboiboi** [reply]
we could also have a structure like:

```python
magpy.defaults.display.backend = 'plotly'
magpy.defaults.itercylinder = -4 # should raise ValueError
magpy.defaults.styles.magnet.magnetization.color.middle = 'white'

etc...

```

**Reply by Alexboiboi** [reply]
implemented in 3798a63d079e8a8d6bca1a739fbeeae5ac54970e to try out and get some feedback here

**Reply by OrtnerMichael** [reply]
I dont have a `magpy.defaults.styles` option

and `magpy.Config` is gone

**Reply by Alexboiboi** [reply]
> I dont have a `magpy.defaults.styles` option
the `style` parameter is under `magpy.defaults.display.styles`

> and `magpy.Config` is gone

`magpy.Config` has become obsolete since it is replaced by `magpy.defaults`  ;)

**Comment by OrtnerMichael** [comment]
# Style Computation overhead

Not to forget:
- [x] style class initialization overhead only when individual styles are set

**Reply by Alexboiboi** [reply]
In order to avoid the style creation overhead, I modified the `style` property getter in the `BaseGeo` class. The property only gets created if it is explicitly set or if it is explicitly called.  This change does not affect user experience and tooltips show up in VsCode even if `style` is not set explicitly.

Here some quick performance comparison:
![image](https://user-images.githubusercontent.com/29252289/139744341-3144dc73-b421-42b4-839e-85a73340a587.png)

implemented here f8b7467c194b2546ffd6e0bbe1ef3b47f8c3ac57

**Reply by OrtnerMichael** [reply]
das funkt auch bei mir !

**Comment by OrtnerMichael** [comment]
# BUG007

Bad magnetization colors of CylinderSegment

```
import magpylib as magpy
src = magpy.magnet.CylinderSegment((0,0,1), (1,2,.2,0,90))
src.display(backend='plotly')
```

![image](https://user-images.githubusercontent.com/47384012/140510600-9b201d7f-6a13-4309-a339-aa5e9298ed3f.png)

**Reply by Alexboiboi** [reply]
The magnetization coloring is computed by giving each vertex a magnitude value ranging from 0 to 1.  The way it is computed in magpylib, is by calculating the normalized distance of a vertex from the plane that is normal to the magnetization vector and passes by the barycenter of the vertices.
The colorscale then is applied by the plotly with this values by the plotly library on its own.

Here you see an edge case where the rendering is not so pretty. If you have a better way to do it, I am al ears, but so far I don't see how else to compute it.

see here how it is implemented for now
https://github.com/magpylib/magpylib/blob/plotly_dev/magpylib/_lib/display/plotly_draw.py#L90-L117

**Comment by OrtnerMichael** [comment]
# Naming discussion
- [x] `Collection.set_styles()`
- [x] `magpy.defaults`
- [x] `animate_time`
- [x] `display.style` instead of `display.styles`
- [x]  object families without `'s'` (magnet, current, dipole, sensor, marker)
    - [x] `markers` has still the `s`
- [x]  `current.arrow`

# Must be done
- [x] docstrings `display()`, `collection.update_styles()` including style input format and examples @OrtnerMichael 
- [x] display magic styles unknown commands do not generate an error msg, e.g. `src.display(style_w00t=13)`
- [x] fix collection.set_styles(dict)

# Missing features to issues 
- Paramviz code reduction
- code inspector does not find options like `style.magnetization...` so they do not appear in VSC tool tips.
- bring coverage to 100 %
- closing test fig
- style doku
- setup.py dev install

# multi-functional parameter discussions
- [x] `style.magnetization.color.middle= color / False / 'auto'`
    - new parameter `style.magnetization.color.mode  = 'tricolor'` options `'bicolor'`, `'tricolor'`, `'tricycle'`
    - and only `style.magnetization.color.middle= standard_grau`

# Other
- [x] display projection plotly original
- [x] display view angle in accordance with Matplotlib
- [x] config -> defaults_class

**Reply by Alexboiboi** [reply]
>multi-functional parameter discussions

I think it is going to be hard to modify the `path` argument behavior without introducing many additional parameter, which may bring confusion to the end user, but for the `magnetization.color.middle` I think It is doable

**Reply by Alexboiboi** [reply]
> * display magic styles unknown commands do not generate an error msg

implemented in c7fb148ec898fa71d3a12797d55c72c198998ac4

**Reply by Alexboiboi** [reply]
> * `magpy.config` or `magpy.defaults` or `magpy.settings`

I am still in favor of keeping `defaults`, since the name explicitly says that these are only "default values" which may be overridden at multiple levels in the code before anything is displayed.

**Reply by OrtnerMichael** [reply]
my proposal
- view angle = Matplotlib
- projection Plotly native

**Comment by OrtnerMichael** [comment]
# BUG 008

```
import magpylib as magpy
src = magpy.magnet.Sphere(magnetization=(0,0,1), diameter=1)
src.move([(.1,0,0)]*50, increment=True)
magpy.defaults.display.styles.markers.marker.size = 10
magpy.display(src)
magpy.display(src, backend='plotly')
```

Plotly doesn't show visible path markers by default.

![image](https://user-images.githubusercontent.com/47384012/140610082-4b94e7f4-d1f1-433c-ae16-8e68df8de478.png)

![image](https://user-images.githubusercontent.com/47384012/140610095-5a02b35d-171d-452e-b04c-cd425cf862b8.png)

**Reply by Alexboiboi** [reply]
This is not a bug ;)

In your code example you are setting the properties of the `Markers` object which defines the style of the `markers` arguments in the `display` function.

The path properties are in `magpy.defaults.display.styles.base.path`

I modified your example with setting defaults for `path.marker` and `markers.marker`

```python
import magpylib as magpy
src = magpy.magnet.Sphere(magnetization=(0,0,1), diameter=1)
src.move([(1,0,0)]*5, increment=True)
magpy.defaults.display.styles.base.path.marker.size = 10
magpy.defaults.display.styles.markers.marker.size = 20
markers = [(0,0,1),(1,0,1)]
magpy.display(src, markers=markers)
magpy.display(src, markers=markers, backend='plotly')
```
<img src="https://user-images.githubusercontent.com/29252289/140617776-1be8ade1-2b03-4408-98dc-20aac9552b96.png" width="300">

**Comment by OrtnerMichael** [comment]
# Question

1. Do I have to unpack a style-dict when passed to `Collection.set_styles()` method ?
2. also it does not catch bad underscore_magic inputs

```
import magpylib as magpy
src1 = magpy.magnet.Sphere((1,1,1), 1)
src2 = magpy.current.Circular(1, 1, (1,0,0))
col = magpy.Collection(src1, src2)

# how to do this ? seems more logic
style_dict = {'color':'r'}
col.set_styles(style_dict)

# this does not generate an error msg
col.set_styles(woot=3)
col.display()
```

**Comment by OrtnerMichael** [comment]
# animate

```
import magpylib as magpy

n = 300
T = 15

src = magpy.magnet.Cuboid((1,2,3), (1,2,3))
src.rotate_from_angax([3]*n, 'z', increment=True)
magpy.display(src,
    path='animate',
    backend='plotly',
    animate_time=T)
```

- [x] change annotation "frame" -> "path index"
- [x] bad computation of number of  frames in warning
- [x] stop print of fps
- [x] add fps to figure
